---
title: "AD Target Gene Report"
author: "Jake Gockley"
date: "11/11/2019"
header-includes:
  - \usepackage{multicol}
  - \newcommand{\btwocol}{\begin{multicols}{2}}
  - \newcommand{\etwocol}{\end{multicols}}
output:
  html_document: default
  pdf_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
library(readr)
library(biomaRt)
library(knitr)
library(kableExtra)
library(tidyverse)
library(grid)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(gridExtra)
library(yaml)
knitr::opts_knit$set(root.dir = config$filedir )
print(getwd())
#webshot::install_phantomjs()
```

```{r yamlIMPT, eval = T, include=T}
print(getwd())
#Assign your YAML formated config file to a varable
configuration <- "YAML_FILE"
config <- read_yaml(file=paste0('./',configuration))

setwd(config$filedir)
plotdir=paste0( 'runs/', config$runname, '/figures' )
tabledir=paste0( 'runs/', config$runname, '/tables' )

Tissues <- config$tissue

#Create Synapse Run Folder and retrieve parentID
CODE <- syn_temp$store(synapseclient$Folder(name = config$runID, parentId = config$parentID))

foo <- syn_temp$getChildren('syn21532475')
Temp <- reticulate::iterate(foo, simplify = F)
for( i in 1:length(Temp) ){
  if( Temp[[i]]$name == config$runID ){
    RunParent <- Temp[[i]]$id
  }else{
  }
}


#DE  = 'syn14237651'
#eQTL_Cer = 'syn16984411'
#eQTL_TCX = 'syn16984410'
#eQTL_DLPFC = 'syn16984409'
#eQTL_EasyYesNo = 'syn12514912
#AgoraDrugability = 'syn13363443'
#Proteomics = 'syn18689335'
#AD Associated Mets Networks: syn11932957
#AD Meta Network Modual Associations syn11954640
```

Importing genelist `r config$genelistfile` as `r config$genelisttype`
```{r GeneNames, eval=T, include=T}

GeneData <- function( id, type ){
  #Retreives gene name info
  #'@id the gene ids to source a list from
  #'@type either ENSG or GNAME for type of gen ids
  
  org = 'hsa'
  #host = 'grch38.ensembl.org'
  #"ensembl",dataset="hsapiens_gene_ensembl")
  #ensembl <- useMart("ENSEMBL_MART_ENSEMBL", host = host)
  ensembl <- useMart("ensembl",dataset="hsapiens_gene_ensembl")
  ds <- listDatasets(ensembl)[, "dataset"]
  ds <- grep(paste0("^", org), ds, value = TRUE)
  
  if (length(ds) == 0){
    stop(paste("Mart not found for:", org))
  } else if (length(ds) > 1) {
    message("Found several marts")
    sapply(ds, function(d) message(paste(which(ds == d), d, sep = ": ")))
    n <- readline(paste0("Choose mart (1-", length(ds), ") : "))
    ds <- ds[as.integer(n)]
  }
    
  if(type == 'ENSG'){
    id.type = 'ensembl_gene_id'
    ensembl <- useDataset(ds, mart = ensembl)
    Genes <- getBM(filters = id.type, attributes = c(id.type, 'hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'), values = id, mart = ensembl)
  }else{
    if(type == 'GNAME'){
      id.type = 'hgnc_symbol'
      ensembl <- useDataset(ds, mart = ensembl)
      Genes <- getBM(filters = id.type, attributes = c('ensembl_gene_id', id.type, 'chromosome_name', 'start_position', 'end_position'), values = id, mart = ensembl)
      ##Cut Decmals
    }else{
      #Error Gene names not specififed correctly
      stop("ERROR: SOURCE=Config.yaml Issue=genelisttype must be ENSG or GNAME")
    }
  }
  return(Genes)
}

#Load Gene Set:
Genes <- read.table( config$genelistfile , header=F, sep='\t')

Tab <- GeneData( Genes, config$genelisttype )
#Trans used as raw for module 7
Trans <- Tab
Tab$Coord_hg19 <- paste0("chr", Tab$chromosome_name, ":", Tab$start_position, "-",Tab$end_position)
Tab$Interval_hg19 <- NA

for( i in 1:dim(Tab)[1]){
  if(  Tab$start_position[i] < Tab$ensembl_gene_id[i] ){
    Tab$Interval_hg19[i] <- paste0("chr", Tab$chromosome_name[i], ":", as.numeric(Tab$start_position[i])-1e6, "-", as.numeric(Tab$end_position[i])+1e6)
  }else{
    Tab$Interval_hg19[i] <- paste0("chr", Tab$chromosome_name[i], ":", as.numeric(Tab$start_position[i])+1e6, "-", as.numeric(Tab$end_position[i])-1e6)
  }
}
```

```{r Module1, eval = T, include=T}
if( 1 %in% config$modules ){
  writeLines(paste0('Modeling and ranking genelist ', config$genelistfile ,' by IGAP SNP evidence')) 
  
  source('code/modules/01-IGAP_module.R')
  
}else{
  writeLines( paste0('Skipping IGAP based Model and Gene-Ranking: User specified module 1 not included in ', config$name) )
}

```

```{r Module2, eval = T, include=T}
if( 2 %in% config$modules ){
  writeLines(paste0( 'Modeling and ranking genelist ', config$genelistfile ,' by Differential Expression evidence'))
  
  source('code/modules/02-DiffExpression_module.R')
  
}else{
  writeLines( paste0('Skipping Differential Expression based Model and Gene-Ranking: User specified module 2 not included in ', config$name) )
}

```

```{r Module3, eval = T, include=T}
if( 3 %in% config$modules ){
  writeLines(paste0('Modeling and ranking genelist ', config$genelistfile ,' by Differential Proteomic Expression evidence'))
  
  source('code/modules/03-DiffProt_module.R')
  
}else{
  writeLines( paste0('Skipping Differential Proteomic Expression based Model and Gene-Ranking: User specified module 3 not included in ', config$name) )
}

```

```{r Module4, eval = T, include=T}
if( 4 %in% config$modules ){
  writeLines(paste0( 'Modeling and ranking genelist ', config$genelistfile ,' by eQTL evidence') )
  
  source('code/modules/04-eQTL_module.R')
  
}else{
  writeLines( paste0('Skipping eQTL based Model and Gene-Ranking: User specified module 4 not included in ', config$name) )
}

```

```{r Module5, eval = T, include=T}
if( 5 %in% config$modules ){
  writeLines(paste0( 'Modeling and ranking genelist ', config$genelistfile ,' by Coexpression Module evidence') )
  
  source('code/modules/05-CoExp_module.R')
  
}else{
  writeLines( paste0('Skipping Coexpression Module based Model and Gene-Clustering: User specified module 5 not included in ', config$name) )
}

```

```{r Module6, eval = T, include=T}
if( 6 %in% config$modules ){
  writeLines(paste0( 'Modeling and ranking genelist ', config$genelistfile ,' by Module Cell-Type Enrichments evidence'))
  
  source('code/modules/06-CellType_module.R')
  
}else{
  writeLines( paste0('Skipping Coexpression Module based Model and Gene-Clustering: User specified module 6 not included in ', config$name) )
}

```

```{r Module7, eval = T, include=T}
setwd( config$filedir )
if( 7 %in% config$modules ){
  writeLines(paste0('Modeling and ranking genelist ', config$genelistfile ,' by Differential Expression evidence'))
  
  source('code/modules/07-Coexpression.R')
  
}else{
  writeLines( paste0('Skipping Partial Coexpression Analysis: User specified module 7 not included in ', config$name) )
}

```