---
title: "AD Target Gene Report"
author: "Jake Gockley"
date: "11/11/2019"
header-includes:
  - \usepackage{multicol}
  - \newcommand{\btwocol}{\begin{multicols}{2}}
  - \newcommand{\etwocol}{\end{multicols}}
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
library(readr)
library(biomaRt)
library(knitr)
library(kableExtra)
library(tidyverse)
library(grid)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(gridExtra)
library(yaml)
```

```{r yamlIMPT, eval = T, include=T}
#Assign your YAML formated config file to a varable
configuration <- "configs/TestConfig.yaml"
config <- read_yaml(configuration)

setwd(config$filedir)
plotdir=paste0( 'runs/', config$runname, '/figures' )
tabledir=paste0( 'runs/', config$runname, '/tables' )

Tissues <- config$tissue

#DE  = 'syn14237651'
#eQTL_Cer = 'syn16984411'
#eQTL_TCX = 'syn16984410'
#eQTL_DLPFC = 'syn16984409'
#eQTL_EasyYesNo = 'syn12514912
#AgoraDrugability = 'syn13363443'
#Proteomics = 'syn18689335'
#AD Associated Mets Networks: syn11932957
#AD Meta Network Modual Associations syn11954640
```

```{r Process, eval=T, include=T}

  writeLines('Importing genelist ', config$genelistfile, ' as ', config$Genelisttype)


```

```{r GeneNames, eval=T, include=T}
GeneData <- function( id ){
  
  org = 'hsa'
  id.type = 'ensembl_gene_id'
  #host = 'jul2019.archive.ensembl.org'
  host = 'grch37.ensembl.org'
  ensembl <- useMart("ENSEMBL_MART_ENSEMBL", host = host)
  ds <- listDatasets(ensembl)[, "dataset"]
  ds <- grep(paste0("^", org), ds, value = TRUE)
  
  if (length(ds) == 0){
    stop(paste("Mart not found for:", org))
  } else if (length(ds) > 1) {
    message("Found several marts")
    sapply(ds, function(d) message(paste(which(ds == d), d, sep = ": ")))
    n <- readline(paste0("Choose mart (1-", length(ds), ") : "))
    ds <- ds[as.integer(n)]
  }
    
  ensembl <- useDataset(ds, mart = ensembl)
  
  Genes <- getBM(filters = id.type, attributes = c(id.type, 'hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'), values = id, mart = ensembl)
  return(Genes)
}

Tab <- GeneData( Genes )
Tab$Coord_hg19 <- paste0("chr", Tab$chromosome_name, ":", Tab$start_position, "-",Tab$end_position)
Tab$Interval_hg19 <- NA
for( i in 1:dim(Tab)[1]){
  if(  Tab$start_position[i] < Tab$ensembl_gene_id[i] ){
    Tab$Interval_hg19[i] <- paste0("chr", Tab$chromosome_name[i], ":", as.numeric(Tab$start_position[i])-1e6, "-", as.numeric(Tab$end_position[i])+1e6)
  }else{
    Tab$Interval_hg19[i] <- paste0("chr", Tab$chromosome_name[i], ":", as.numeric(Tab$start_position[i])+1e6, "-", as.numeric(Tab$end_position[i])-1e6)
  }
}

```
```{r Module1, eval = T, include=T}
if( i %in% config$modules ){
  writeLines('Modeling and ranking genelist ', config$genelistfile ,' by IGAP SNP evidence')
  
  source()
  
}else{
  writeLines('Skipping IGAP based Model and Gene-Ranking: User specified module 1 not included in <Config>.yaml ')
}

```


