---
title: "Meta analysis of reprocessed RNASeq data from AMP-AD for hg38 Reprocessed Data (all 7 brain regions)"
author: "Adapted from Thanneer Perumal's code"
date: "`r date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file
## Load required libraries
library(CovariateAnalysis)
library(data.table)
library(plyr)
library(tidyverse)
library(knitr)
library(githubr)
library(doParallel)
library(foreach)
cl = makeCluster(detectCores()-2)
registerDoParallel(cl)
options(xtable.type="html")
knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
reticulate::use_python("/usr/bin/python", required = TRUE)
synapseclient <- reticulate::import("synapseclient")
syn_temp <- synapseclient$Synapse()
syn_temp$login()

#parentId = 'syn18936948';
parentId = 'syn22351719';
activityName = 'Meta analysis of differential expression hg38 Reprocessed RNASeq';
activityDescription = 'Fixed and random effect meta-analysis of AMP-AD data hg38 Reprocessed Data (6 brain regions - No CBE)';
thisFileName <- 'RNA-Seq_Meta_Analysis.Rmd'
# Github link
thisRepo <- getRepo(repository = "Sage-Bionetworks/AD_TargetRank", ref="branch", refName='master')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('code/',thisFileName))
```

## Data download
### Get study specific covariates and adjusted expression from synapse

```{r get.exp}
# Function to read tsv/csv files from synapse
downloadFile <- function(id){ 
  fread(syn_temp$get(id)$path, data.table = F)
}
# Read covariates
covar.ids = c(ROSMAP = 'syn8456631', MSSM = 'syn8484996', MAYO = 'syn8466814')
all.used.ids = as.character(covar.ids)
covar =  covar.ids %>%
  lapply(downloadFile)

# Read adjusted expression
expr.ids = c(ROSMAP = 'syn21292041', MSSM = 'syn21285564', MAYO = 'syn21291908')
all.used.ids = c(all.used.ids, as.character(expr.ids))
expr =  expr.ids %>%
  lapply(downloadFile) %>%
  plyr::join_all(type = 'inner')
```

```{r modify.covar}
covar$ROSMAP = covar$ROSMAP %>%
  dplyr::select(SampleID, Sex, Diagnosis, APOE4) %>%
  dplyr::mutate(Tissue = 'DLPFC')
covar$MSSM = covar$MSSM %>%
  dplyr::select(SampleID, Sex, Tissue.Diagnosis, Tissue.APOE4) %>%
  dplyr::rename(Sex = Sex) %>%
  tidyr::separate(Tissue.Diagnosis, c('Tissue1', 'Diagnosis'), sep = '\\.') %>%
  tidyr::separate(Tissue.APOE4, c('Tissue', 'APOE4'), sep = '\\.')
covar$MAYO = covar$MAYO %>%
  dplyr::select(SampleID, Sex, Tissue.Diagnosis, Tissue.APOE4) %>%
  tidyr::separate(Tissue.Diagnosis, c('Tissue1', 'Diagnosis'), sep = '\\.') %>%
  tidyr::separate(Tissue.APOE4, c('Tissue', 'APOE4'), sep = '\\.')
covar = covar %>%
  data.table::rbindlist(use.names = T, fill = T, idcol = 'Study') %>%
  dplyr::select(-Tissue1)
```

## Perform fixed and random effect meta-analysis for each gene across 7 different brain regions
#### Meta-analysis of AD-CONTROL
```{r ad.cntrl}
# Mean and sd of gene expression in each tissue x diagnosis group
tissue.dx.summary = plyr::ddply(covar, .(Tissue, Diagnosis), .fun = function(x, y){
  data.frame(ensembl_gene_id = y$ensembl_gene_id,
             n = dim(x)[1],
             mn = rowMeans(y[,x$SampleID], na.rm = T),
             sd = apply(y[,x$SampleID], 1, sd, na.rm = T))
}, expr)
# Perform meta-analysis for AD-CONTROL comparison - Exclude CBE
meta.anlz.ad_cntrl = plyr::ddply(tissue.dx.summary[(tissue.dx.summary$Tissue%in%'CBE')==F,], .(ensembl_gene_id), .fun = function(x){
  exp.effect = dplyr::filter(x, Diagnosis == 'AD')
  rownames(exp.effect) = exp.effect$Tissue
  cntrl.effect = dplyr::filter(x, Diagnosis == 'CONTROL')
  rownames(cntrl.effect) = cntrl.effect$Tissue
  cntrl.effect = cntrl.effect[rownames(exp.effect), ]
  
  tmp = metacont(exp.effect$n, exp.effect$mn, exp.effect$sd, 
                 cntrl.effect$n, cntrl.effect$mn, cntrl.effect$sd,
                 studlab = exp.effect$Tissue,
                 sm = 'SMD', method.smd = 'Hedges',
                 method.tau = 'REML')
  
  return(data.frame(tmp[c('TE.fixed', 'seTE.fixed', 'lower.fixed', 'upper.fixed', 'zval.fixed', 'pval.fixed',
                   'TE.random', 'seTE.random', 'lower.random', 'upper.random', 'zval.random', 'pval.random',
                   'Q', 'tau', 'H', 'I2')]))
}, .parallel = TRUE, .paropts = list(.packages = c('meta', 'dplyr'))) %>%
  dplyr::mutate(fdr.fixed = p.adjust(pval.fixed, method = 'fdr'),
                fdr.random = p.adjust(pval.random, method = 'fdr'))
p = list()
p[[1]] = ggplot(meta.anlz.ad_cntrl, aes(x = -log10(fdr.fixed), y = -log10(fdr.random)))+geom_point()
p[[2]] = ggplot(meta.anlz.ad_cntrl, aes(y = -log10(fdr.fixed), x = TE.fixed))+geom_point()+geom_hline(yintercept = -log10(0.05), color = 'red')
p[[3]] = ggplot(meta.anlz.ad_cntrl, aes(y = -log10(fdr.random), x = TE.random))+geom_point()+geom_hline(yintercept = -log10(0.05), color = 'red')
ggpubr::ggarrange(plotlist = p, ncol = 3, nrow = 1)
# Get gene sets
tmp = meta.anlz.ad_cntrl %>%
  dplyr::select(ensembl_gene_id, TE.fixed, TE.random, fdr.fixed, fdr.random) %>%
  tidyr::gather(Feature, Value, -ensembl_gene_id) %>%
  tidyr::separate(Feature, c('Feature1', 'Model'), sep = '\\.') %>%
  tidyr::spread(Feature1, Value) %>%
  dplyr::filter(fdr <= 0.05, abs(TE) >= 0.2)
tmp$Direction[tmp$TE <= 0] = 'DOWN'
tmp$Direction[tmp$TE >= 0] = 'UP'
gs = tmp %>%
  plyr::dlply(.(Model, Direction), .fun = function(x){
    unique(x$ensembl_gene_id)
  })
names(gs) = paste0('ad.control.', names(gs))
all.gs = gs
```

#### Adjust the AD-Control Model for weights
```{r weights}
meta.anlz.ad_cntrl$TE.random.abs <- abs( meta.anlz.ad_cntrl$TE.random )

#Assign Sig Or No
meta.anlz.ad_cntrl$Sig <- 'NO'
meta.anlz.ad_cntrl[meta.anlz.ad_cntrl$pval.random < 0.05, ]$Sig <- 'YES'

meta.anlz.ad_cntrl <- meta.anlz.ad_cntrl[ order(meta.anlz.ad_cntrl$TE.random.abs), ]
meta.anlz.ad_cntrl$y <- c( 1:dim(meta.anlz.ad_cntrl)[1] )/dim(meta.anlz.ad_cntrl)[1]

meta.anlz.ad_cntrl$Type <- 'Actual Rank'

mylogit <- glm(meta.anlz.ad_cntrl$y  ~ abs( meta.anlz.ad_cntrl$TE.random.abs), data = meta.anlz.ad_cntrl, family = "binomial")

#Make Weight Template
Work <- meta.anlz.ad_cntrl[, c('ensembl_gene_id','TE.random', 'seTE.random','fdr.random','TE.random.abs')]
Work$log_abs_PVal <- abs( meta.anlz.ad_cntrl$TE.random.abs )
colnames(Work)[ colnames(Work) == 'ensembl_gene_id' ] <- 'ENSG'

#Calculate predicted Weights
Work2 <- Work  
Work2$Type <- "Predicted Weight"
#Work2 <- Work2[ , c('ENSG', 'log_abs_PVal',"Type") ]
Work2$y <- 1/(1+exp( -( mylogit$coefficients[1]+mylogit$coefficients[2]* abs( meta.anlz.ad_cntrl$TE.random.abs ) ) ))
Work2$PreAdjWeight <- Work2$y

#Add Actual Ranks to Template
Work$y <- c( 1:dim(Work)[1] )/dim(Work)[1]
Work$Type <- 'Actual Rank'
mylogit <- glm(Work$y  ~ abs( Work$TE.random.abs), data = Work, family = "binomial")
Work$PreAdjWeight <- Work$y

Weights <- as.data.frame( rbind(Work,Work2) )
Weights[ Weights$fdr.random > 0.05,]$y <- 0
Weights <- Weights[ order(-Weights$y), ]

#Annotate with Gene Names
library('biomaRt')
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- as.character(Weights$ENSG[ !duplicated(Weights$ENSG) ])
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=genes,mart= mart)

G_list$ensembl_gene_id <- as.character(G_list$ensembl_gene_id)
G_list$hgnc_symbol <- as.character(G_list$hgnc_symbol)

G_list <- G_list[ !duplicated(G_list$ensembl_gene_id),]
Miss <- cbind( ensembl_gene_id = as.character(Weights[ (Weights$ENSG %in% G_list$ensembl_gene_id) == F,]$ENSG),
               hgnc_symbol = rep("NA", length(as.character(Weights[ (Weights$ENSG %in% G_list$ensembl_gene_id) == F,]$ENSG)) )
               )
G_list <- rbind( Miss, G_list)
G_list <- G_list[ !duplicated(G_list$ensembl_gene_id),]

row.names(G_list) <- G_list$ensembl_gene_id
Weights$Gene <- G_list[ as.character(Weights$ENSG),]$hgnc_symbol

Weights <- Weights[, c( "ENSG", "Gene", "TE.random", "seTE.random", "fdr.random", 'TE.random.abs', "log_abs_PVal", "Type", "PreAdjWeight", "y" )]
```

## Store results in synapse
```{r syn.store, cache = FALSE, include=FALSE}
CODE <- syn_temp$store(synapseclient$Folder(name = "MetaAnalysis", parentId = parentId))
Syns_Used <- c( 'syn8456631', 'syn8484996', 'syn8466814', 'syn21292041', 'syn21285564', 'syn21291908' )

# Write results to files
fwrite(meta.anlz.ad_cntrl, file = 'meta.anlz.ad_cntrl.tsv', sep = '\t', row.names = F, quote = F)
ENRICH_OBJ <-  syn_temp$store( synapseclient$File( path='meta.anlz.ad_cntrl.tsv', name = 'RNASeq AD-Control meta-analysis across 6 brain regions (No CBE) hg38 Reprocessed Data', parentId=CODE$properties$id ), used = Syns_Used, activityName = activityName, executed = thisFile, activityDescription = activityDescription)
```

## Source code
